### PR TITLE
ARROW-4139: [C++][Parquet] Logical Type casting for parquet statistics

### PR DESCRIPTION
**Pull request by mpeaton**
Introductory PR for ARROW-4139

### CODE REVIEW

File: python/pyarrow/parquet.py [hunk 1]
```diff
@@ -1407,3 +1406,52 @@ def read_schema(where, memory_map=False):
...
+            df.loc[i] = [lmin, lmax, 'NONE']
+
+        elif ps.column(i).logical_type == 'TIMESTAMP_MICROS':
+            lmin = datetime64(rg.column(i).statistics.min, 'us')
+            lmax = datetime64(rg.column(i).statistics.max, 'us')
+            df.loc[i] = [lmin, lmax, 'TIMESTAMP_MICROS']
+        else:
+            lmin = rg.column(i).statistics.min
+            lmax = rg.column(i).statistics.max
+            df.loc[i] = [lmin, lmax, 'UNKNOWN']
```
Comment (mrocklin) [inline python/pyarrow/parquet.py]: This seems straightforward, but I have two concerns:

-  Are there other types that we would also need to convert?
-  Is there already some place in the Arrow codebase that does this kind of logic for us,  but in a centralized way that would be guaranteed to track changes in the future?

I'm concerned that this approach would require constant changes in the future, which would increase friction between Arrow and downstream libraries, and would increase maintenance burden in the future.

This was discussed here: https://issues.apache.org/jira/browse/ARROW-4139?focusedCommentId=16823025&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-16823025

**Comment by wesm** [comment]
I'm also concerned about the maintainability of these changes. I'm still catching up from vacation (including a lot of non-engineering Ursa Labs organization stuff...) but I will carve out some time to have a closer look at this and also @xhochy's changes in https://github.com/apache/arrow/pull/2623/files. 

Regarding that patch, I want to be careful about introducing unnecessary new public API surface area since we have been discussing porting much of this functionality into C++ so that it can be used in other languages and not just Python (see https://docs.google.com/document/d/1bVhzifD38qDypnSjtf8exvpP3sSB5x_Kw9m-n66FB2c/edit?usp=sharing). Of course, I'm sympathetic to the requirements that you have for Dask so I would like to unblock the use cases you have in the meantime

**Comment by mpeaton** [comment]
I'll revisit  https://github.com/apache/arrow/pull/2623/files and try to deduce how to perform type checking and conversion for UTF8 and datetime64 logical types.  Thanks.

**Comment by mrocklin** [comment]
Just to be clear, the objective of the original issue wasn't just for UTF and datetime types.  It was to handle conversion for all types that require conversion, both for current and future types.

**Comment by wesm** [comment]
From an API perspective it seems like a solution would be to add `min_as_logical` and `max_as_logical` attributes (or similar) to `RowGroupStatistics` here

https://github.com/apache/arrow/blob/master/python/pyarrow/_parquet.pyx#L87

I don't think collecting the statistics as a particular pandas DataFrame arrangement is necessarily something we want to maintain, you could have that code in Dask

**Comment by mrocklin** [comment]
I'm fine with the current API of getting things.  This isn't likely to be something that end users will care about so we don't need to make this particularly magical.

I might suggest that the default would be logical types rather than physical, but I don't really care either way.  I'm happy with any API as long as I can get the information robustly.

**Comment by mpeaton** [comment]
As I understand it, statistics should generally be stored in Parquet, for types that allow it, though I am still having trouble determining how best to access them in an efficient way.  According to this Cloudera reference https://blog.cloudera.com/blog/2017/12/faster-performance-for-selective-queries/.  Also, does anyone have any tips on establishing a debug environment using docker-compose or other?

**Comment by wesm** [comment]
What kind of debug environment are you looking to create?

**Comment by mpeaton** [comment]
My primary use case is development for this task so really cython debugging of parquet.so using lldb.  I have thus far had trouble setting breakpoints and assumed that maybe the debug builds are not including symbols when $ARROW_BUILD_TYPE is set to debug?...      As a tangent, I noted that some of the cpp tests which appeared related such as test-statistics would fail on my local build, so I also attempted to use docker-compose images running in interactive mode for a complementary build path to check both debugging build symbol presence, as well as cpp test status, but found that docker-compose up eventually failed on my machine.  From my investigation of the docker images in my registry, it appears that they do not have arrow built inside ( or more specifically debug builds ) so overall I am wondering if there is a secret manual I might be missing, and if there are any best practices available that may have been circulated amongst the community.  Thanks for responding @wesm I don't really want to be a burden I'm trying to help here :).

**Comment by wesm** [comment]
I have no experience debugging the Cython extensions so I'm not sure what to advise

**Comment by mpeaton** [comment]
I expect that this PR does not quite address JIRA issue 4139 from a performance aspect, and that the desired implementation will require a more sophisticated change.  Of course I am interested in continuing to pursue that.  Has there been any additional related work on how to best handle statistics and metadata, particularly as the pertain to providing the utility required by @mrocklin , and thus I presume Dask?

**Comment by wesm** [comment]
I haven't been able to take a look at this since last writing unfortunately

**Comment by wesm** [comment]
I claimed the issue for the Arrow 0.14.0 release. I'll close this PR for now and return to it when I work on the issue (unless someone else beats me to it)

**Pull request closed without merging**